### DISCUSSION TITLE
Fluxcd github secret configuration manual boostrap

### DISCUSSION THREAD

**Discussion by cello86** [fluxcd/flux2#4184] [General] [answered]
Hi all,
I'm trying to configurare flux2 on a GKE cluster and flux2 has been installed via helm chart. I would use an helm chart to define the git repo and helm repo config for the flux pipeline, because I would manage this config via terraform and not via flux cli.

Is it possible to link a secret with the github token to the git repo definition via manifest?

Thanks,
Marcello

**Comment by kingdonb** [comment, marked as answer]
Hi Marcello, the Flux installed as Helm chart is not a configuration that's supported by Flux officially. There are a group of people that only use Helm to install things on their clusters, they might also be using Terraform to help explain why they needed this, but it has a list of quirks that make it undesirable to use and difficult to support.

I think you're asking if it's possible to embed a secret in this configuration, and I think the answer is no. Secret decryption is done by Kustomize controller, so there won't be anything to apply the secret until decryption is configured. You can use KMS secrets but the bootstrap configuration in Flux doesn't support decryption in the root resource. You need to add a separate Flux kustomization and enable decryption there. I think you're trying to seed the cluster with a secret to access the git repo, which is part of bootstrap. This is one of the reasons we made bootstrap, so users wouldn't have to manage this configuration by themselves (it is complicated.)

But I'm not sure this is exactly what you're asking. You may wish to investigate the terraform provider flux, which is supported:

https://github.com/fluxcd/terraform-provider-flux

This can be used to actually bootstrap your clusters, so that Flux is configured the same as it would be on any cluster that has been bootstrapped using the normal instructions, only via Terraform. Then Flux can upgrade itself without Terraform involvement, Flux manages all the resources on your cluster, and Helm isn't involved. Flux is better at managing Kubernetes resources than the Hashi terraform providers for Kubernetes and Helm, it doesn't have the problem where it can go out of sync with the actual cluster state when a cluster is destroyed and re-created for example – Flux is idempotent so just bootstrap again. The new provider is called the bootstrap provider, it implements the same logic as Flux's CLI for bootstrapping. It's up to the GitHub provider to configure the deploy key, there are further examples for both GitHub and GitLab that show how to do this configuration from end-to-end.

This way you can link your github secret (say you have created a scoped access token for the repo) to the repository, or set it up as an SSH deploy key and use the ssh remote. The main difference between this and Flux's normal bootstrap is that the bootstrap provider does not provide the secret generation itself, (I think it's expected to get that secret from a vault or generate it somehow else in Terraform, or externally manage it somehow else.)

**Reply by cello86** [reply]
@kingdonb thanks for the detailed response. Ideally on the provider page is reported https://registry.terraform.io/providers/fluxcd/flux/latest/docs/guides/install-helm-release if you don't need to boostrap flux and the repository and this permit to have only the basic components for flux arhitecture.

The second step is to define the best approach to put under terraform the single deployments managed by flux. In fact you can:

execute flu-cli command via local-exec provider
add kubernetes manifests for secret, helmrelease and gitrepository and applied by kubernetes provider
add kubernetes manifests for secret, helmrelease and gitrepository packaged into an helm chart config and applied by helm provider
All the options provide the same final result but there are differences and prons/cons to identify.

Marcello
